### PR TITLE
Proxy some routes instead of redirecting

### PR DESCRIPTION
**Pull request by betatim**
This changes the federation redirector so that most pages a user interacts with are proxied to gke.mybinder.org instead of redirecting people.

This means that as a user you see mybinder.org in the address bar when you visit the homepage and when waiting for a repo to launch. After this you are redirected.

You can try it out locally by starting `app.py` and then visiting http://localhost:8080/. What do people think of this change of behaviour? I think it makes it harder for people to copy&paste the "wrong" URL and with the new banner, the subdomains working and users seeing the hoster's name in the redirected URL we have a better balance between usability and exposing who is donating compute resources.

### CODE REVIEW

**Comment by minrk** [comment]
Seems like it accomplishes the right thing. I'm still thinking about the implementation. What happens if a user refreshes the loading page? Will they get routed to a new cluster, or is there anything sticky that keeps them on the given cluster? What about sequential requests, e.g. a reconnecting eventstream? Will they reliably go to the same host?

**Comment by betatim** [comment]
The redirect handler uses a cookie to make things "sticky" but that cookie only lasts for 10s. It gets set each time you make a request so if you make a (relevant) request every 8s you will always get the same cluster.

I picked 10s because it seemed like a good starting guess to balance "distribute people" and "keep people who hit refresh a bunch on the same connection".

Does a event stream reconnect? I thought that at the moment when I visit https://gke.mybinder.org/v2/gh/binder-examples/requirements/master (so no proxy, straight up binderhub) and press refresh I trigger a new launch and that the same thing would happen if my connection dropped for some reason.

I was thinking of working on https://github.com/jupyterhub/mybinder.org-deploy/issues/1066 (the title is wrong, should mention the federation) a bit more and through that solving the issue of sticky-ness. It would route you to a cluster based on which repo you are launching (modulo hot repos that somehow get distributed across clusters).

**Comment by betatim** [comment]
Idea from the weekend: use a cookie per path (`/build/<spec>`) to remember which cluster we assigned the user to. This means we can use a really long (or infinite?) timeout as long as we check that the cluster is currently available. We still balance users across clusters. Need to check that the `path` attribute of cookies works as I think it does (cookie only gets sent if a request is for the (sub)path set in the cookie.

File: images/federation-redirect/app.py [hunk 1]
```diff
@@ -60,7 +111,7 @@ def set_default_headers(self):
         # make sure the host is a valid choice and considered healthy
         if host_name not in self.host_names:
             host_name = random.choices(self.host_names, self.host_weights)[0]
-        self.set_cookie("host", host_name, max_age=10)
+        self.set_cookie("host", host_name, path=uri)
```
Comment (betatim) [inline images/federation-redirect/app.py]: By setting the `path` parameter this cookie will only be sent back by the browser if a user visits the same path. So for repeated calls to http://example.com/v2/build/binder-examples/requirements/master we will see this cookie again and send the user to the same cluster again.

However a request to http://example.com/v2/build/binder-examples/r/master will not send the `host` cookie so we will pick a new cluster, which gets stored in a cookie again.

**Comment by betatim** [comment]
@minrk what do you think of the idea of using a path specific cookie? I think it solves the "stickyness" problem.

It would be even nicer if we could do this on a repo basis (ignoring the branch/SHA1/etc) but I don't see a way to do that without having to put logic that understands the repo provider spec for each repo provider into the proxy.

**Comment by betatim** [comment]
I'll merge this today as I'll have some time to watch what happens and revert it.

**Comment by minrk** [comment]
I think we should take this into consideration when designing API v3. Routing cookies are definitely a standard way to do this, but I think we might be able to get away from that if we bake it into the multi-request build/launch API.

**Pull request merged**